### DISCUSSION TITLE
Set Command line arguments without DTE

### DISCUSSION THREAD

**Discussion by simfly80** [VsixCommunity/Community.VisualStudio.Toolkit#413] [Q&A]
How can I set the Command line arguments for a Project without using DTE?

**Comment by RobertvanderHulst** [comment]
In a C# project you can set the ProjectFile property StartArguments. Would that work?
If you have the project object from the toolkit then there is a method TrySetAttributeAsync() that allows you to set the value.

**Comment by simfly80** [comment]
In VS2019 I set the Command line arguments with DTE like this:

EnvDTE.Project _project = (EnvDTE.Project) _projectUIHierarchyItem.Object;
_project.ConfigurationManager.ActiveConfiguration.Properties.Item("StartArguments").Value = "/startconfigwizard /serviceMode";

The information is stored in .csproj.user. In Visual Studio 2019 it looked like this:
![image](https://user-images.githubusercontent.com/124884053/218974827-c48da4c4-d27d-4b4d-9041-64184df0ee51.png)

In VS2022 the upper solution via DTE does not work. When I set the StartArugments via TrySetAttributeAsync() they are only added to the .csproj.  In Visual Studio 2022 they are not shown in the properties and the launchSettings.json does not contain command line arguments. 

![image](https://user-images.githubusercontent.com/124884053/218977145-4147c0c1-d956-4185-abb6-09643520bcd9.png)

![image](https://user-images.githubusercontent.com/124884053/218979976-e4f29ad6-f7f4-446e-b215-ca6c489d64be.png)

Is there a way to edit the Launch Profiles via the toolkit?

**Reply by jamesc-skyward** [reply]
Are you sure this is a VS2019 vs VS2022 thing? I have some similar code that runs in the absence of a launchSettings.json file. I haven't tested it recently, but as far as I know it worked in VS2022. It just stopped being useful after we started updating our .NET Framework projects to the newer SDK-style format. For those we add the start arguments to the .json file.

While making that transition I did run into a bug where the StartArguments would be passed in twice. This ended up being from VS pulling the arguments from _both_ the .csproj.user file and the launchSettings.json file. My fix was to wipe the values out of the .csproj.user file. If you don't want to update the launchSettings.json file, you could probably update the .csproj.user file directly. It's just XML.

Here's how I get the StartArguments:
```
XmlNamespaceManager namespaceManager = new XmlNamespaceManager(xmlDocument.NameTable);
namespaceManager.AddNamespace("t", "http://schemas.microsoft.com/developer/msbuild/2003");
XmlNode xmlNode = xmlDocument.SelectSingleNode("./t:Project/t:PropertyGroup/t:StartArguments", namespaceManager);
```

**Reply by RobertvanderHulst** [reply]
James,
This is not really a VS2019/VS2022 thing. This has to do with the project format.
SDK style projects store the debug settings in the json file. Legacy projects store it in the project file or the user file.

**Comment by RobertvanderHulst** [comment]
> Is there a way to edit the Launch Profiles via the toolkit?

Not that I am aware of. Remember, this is a community driven project. Apparently, nobody has had the need for this.
If you managed to do this, then please add the code to the toolkit and send in a pull request, so Mads can merge it into the toolkit.

**Comment by RobertvanderHulst** [comment]
You can find the code that is reponsible for this in the C# and VB project system here
https://github.com/dotnet/project-system/tree/main/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug
